src/lib/temple/activity-new/utils.ts

```diff
@@ -0,0 +1,207 @@
+      ...activityOperBase,
+      type: 'transaction',
+      destination: operation.target
+    };
+    if (contractAddress != null) activityOper.contractAddress = contractAddress;
+    if (tokenId != null) activityOper.tokenId = tokenId;
+    if (isTzktOperParam(operation.parameter)) activityOper.entrypoint = operation.parameter.entrypoint;
+    return activityOper;
+  }
+
+  const parameter = operation.parameter;
+
+  if (parameter == null) {
+    if (operation.target.address !== address && operation.sender.address !== address) return null;
+
+    const source = operation.sender;
+    const amount = String(operation.amount);
+
+    return _buildReturn({ amount, source });
+  } else if (isTzktOperParam_Fa2(parameter)) {
+    const values = reduceParameterFa2Values(parameter.value, address);
+    const firstVal = values[0];
+    // (!) Here we abandon other but 1st non-zero-amount values
+    // (?) Is this right?
+    if (firstVal == null) return null;
+
+    const contractAddress = operation.target.address;
+    const amount = firstVal.amount;
+    const tokenId = firstVal.tokenId;
+    const source = firstVal.from === address ? { ...operation.sender, address } : operation.sender;
+
+    return _buildReturn({ amount, source, contractAddress, tokenId });
+  } else if (isTzktOperParam_Fa12(parameter)) {
+    if (parameter.entrypoint === 'approve') return null;
+
+    const source = { ...operation.sender };
+    if (parameter.value.from === address) source.address = address;
+    else if (parameter.value.to === address) source.address = parameter.value.from;
+    else return null;
+
+    const contractAddress = operation.target.address;
+    const amount = parameter.value.value;
+
+    return _buildReturn({ amount, source, contractAddress });
+  } else if (isTzktOperParam_LiquidityBaking(parameter)) {
+    const source = operation.sender;
+    const contractAddress = operation.target.address;
+    const amount = parameter.value.quantity;
+
+    return _buildReturn({ amount, source, contractAddress });
+  } else {
+    const source = operation.sender;
+    const amount = String(operation.amount);
+
+    return _buildReturn({ amount, source });
+  }
+}
+
+function buildActivityOperBase(operation: TzktOperation, address: string, amount: string, source: ActivityMember) {
+  const { id, hash, level } = operation;
+  const reducedOperation: ActivityOperationBase = {
+    id,
+    hash,
+    level,
+    source,
+    amountSigned: source.address === address ? `-${amount}` : amount,
+    status: stringToActivityStatus(operation.status),
+    addedAt: operation.timestamp,
+    timestamp: new Date(operation.timestamp).getTime()
+  };
+  return reducedOperation;
+}
+
+function reduceParameterFa2Values(values: ParameterFa2['value'], relAddress: string) {
+  const result: {
+    from: string;
+    amount: string;
+    tokenId: string;
+  }[] = [];
+
+  for (const val of values) {
+    // (!) We assume, that all `val.txs` items have same `token_id`
```

Comment:
visit [https://tezos.b9lab.com/fa2](https://tezos.b9lab.com/fa2), there are link to code in smartpy ide - fa2 token-standard/smartcontract literally have it inside its code